### ISSUE TITLE
Scala plugin + annotation processor

### ISSUE THREAD

**Issue by cmelchior** [realm/realm-java#1531]
Copy from: https://github.com/realm/realm-java/issues/644#issuecomment-145507390

-----
@aafa

I'm having similar problem running realm with SBT + android-plugin.

What I have
1. Annotated scala class that extends RealmObject
2. javacOptions ++= Seq("-processor", "io.realm.processor.RealmProcessor", "-proc:only") That I added just in case
3. compileOrder := CompileOrder.JavaThenScala

What I see
1. META-INF/services/javax.annotation.processing.Processor 
2. But at runtime I got

java.lang.IllegalArgumentException: Class is not part of the schema for this Realm
Important - the problem is only with scala classes, with java classes it works fine

I'm not very familiar with annotation processing, what should I look for to get this working?

**Comment by cmelchior** [comment]
I am not familiar with the Scala plugin, but these answers seem to indicate it is not supported by the Scala compiler:
http://stackoverflow.com/questions/7454018/using-androidannotations-with-scala-and-gradle
https://groups.google.com/forum/#!topic/scala-internals/_IV46CRHsX8

If that is the case, I am afraid you cannot use Scala model classes with Realm as we depend on the Annotation processor to generate code. As you point out, the normal Java classes work, so it seems to be a problem with the Scala compiler :(

**Comment by aafa** [comment]
However, according to this example https://github.com/retronym/annotation-processor java annotation processing + scala classes = should work

I will try to verify this case myself and post results

**Comment by cmelchior** [comment]
Thank you. The answers I found where also rather old, so things might have changed.

**Comment by pfn** [comment]
Scala does not support jsr269 processing. A post compile step must be run to process class files compiled by scala. 

It should be enough to run `javac -processor ... classes-dir`

**Comment by aafa** [comment]
So I was experimenting with class files postprocessing (thanks @pfn for starting point)

With simple model
```
class Dto extends RealmObject {
  @PrimaryKey var id: Long = 1
  var name: String = "test"
}
```
I see this picture
```
Round 1:
        input files: {com.sbtmult.core.db.Dto}
        annotations: [io.realm.annotations.RealmClass, scala.reflect.ScalaSignature, io.realm.annotations.PrimaryKey]
        last round: false
Note: Processing class Dto
error: Only getters and setters should be defined in model classes
Round 2:
        input files: {}
        annotations: []
        last round: true
Note: Creating DefaultRealmModule
warning: File for type 'io.realm.DefaultRealmModule' created in the last round will not be subject to annotation processing.
warning: File for type 'io.realm.DefaultRealmModuleMediator' created in the last round will not be subject to annotation processing.
1 error
2 warnings
```

Model bytecode looks like this
```
public class com.sbtmult.core.db.Dto extends io.realm.RealmObject {
  public long id();
  public void id_$eq(long);
  public java.lang.String name();
  public void name_$eq(java.lang.String);
  public com.sbtmult.core.db.Dto();
}
```

Should I see that error for a bytecode like this?

**Comment by cmelchior** [comment]
Hi @aafa 
Sorry for the late reply. It seems that the Scala compiler adds some methods we don't currently support.

You can see there is `id()`, `id_$eq()`, `name()` and `name_$eq()`. This unfortunately conflict with the current Realm restriction of only having getters and setters. So right now I am afraid your will have to create your model classes in Java.

Note this restriction will be lifted once https://github.com/realm/realm-java/pull/1216 is completed.

**Comment by aafa** [comment]
@cmelchior Ok, thanks!  I'll try again when that issue is resolved.

**Comment by cmelchior** [comment]
No problem, #909 should be resolved really soon which will resolve this. Closing this now.

**Comment by RichardBradley** [comment]
Hi, @pfn -- you said

> Scala does not support jsr269 processing. A post compile step must be run to process class files compiled by scala.
>
> It should be enough to run javac -processor ... classes-dir

I can't get this to work in a Scala-only project, as "javac" requires Java source files as one of its arguments. Do you have an example where this can be used to post-process Scala bytecode?

**Comment by aafa** [comment]
Hi, @RichardBradley

> "javac" requires Java source files

Actually, for annotation processing you can pass already compiled .class 

I used this command 
```
val command = "javac -classpath " + cp + " -processor io.realm.processor.RealmProcessor -proc:only -XprintRounds -s " + classDir + " com.sbtmult.core.db.Dto"
```

Where `cp` is a classpath collection, `classDir` is a dir where my class is located and
`com.sbtmult.core.db.Dto` is a full class name.

**Comment by RichardBradley** [comment]
Thanks, that nearly works! It gets past the "`javac: no source files`" error I was getting from `javac`.

Unfortunately, the annotation processor I'm trying to use doesn't seem to work, as it doesn't recognise the class.

I want to use [Google AutoService](https://github.com/google/auto/tree/master/service), but its processor calls

    roundEnv.getElementsAnnotatedWith(AutoService.class)

which doesn't seem to return the class mirror if it's a Scala class; it returns an empty list.

I have uploaded a demo to https://gist.github.com/RichardBradley/48a3bf13979bafd2c4b7

This is re http://stackoverflow.com/questions/33281999/automatic-meta-inf-services-generation-in-scala-and-sbt-for-serviceloader

**Comment by aafa** [comment]
Hi @cmelchior,  

Since #909 was resolved, I decided to reiterate on this issue.  And everything went very smooth at the beginning. 
1. annotation processing went fine
2. dex step ok
3. bunch of realm generated classes are well packed in apk

But then in a runtime I got this
```
  Caused by: java.lang.ClassCastException: com.github.aafa.model.User cannot be cast to io.realm.UserRealmProxyInterface
            at io.realm.UserRealmProxy.copy(UserRealmProxy.java:250)
            at io.realm.UserRealmProxy.copyOrUpdate(UserRealmProxy.java:245)
            at io.realm.DefaultRealmModuleMediator.copyOrUpdate(DefaultRealmModuleMediator.java:98)
            at io.realm.Realm.copyOrUpdate(Realm.java:1395)
            at io.realm.Realm.copyToRealm(Realm.java:746)
            at com.github.aafa.activity.MainActivity.realmTest(MainActivity.scala:40)
            at com.github.aafa.activity.MainActivity.onCreate(MainActivity.scala:33)
```

Which happening at this point
```
    public static User copy(Realm realm, User newObject, boolean update, Map<RealmObject,RealmObjectProxy> cache) {
        User realmObject = realm.createObject(User.class, ((UserRealmProxyInterface) newObject).realmGet$id());
        cache.put(newObject, (RealmObjectProxy) realmObject);
        ((UserRealmProxyInterface) realmObject).realmSet$id(((UserRealmProxyInterface)newObject).realmGet$id());
        ((UserRealmProxyInterface) realmObject).realmSet$name(((UserRealmProxyInterface)newObject).realmGet$name());
        return realmObject;
    }
```

Indeed User.class is not implementing UserRealmProxyInterface

```
package com.github.aafa.model;
import io.realm.RealmObject;
public class User extends RealmObject {
...
}
```


What may be causing this? Something went wrong with code generation?

**Comment by cmelchior** [comment]
I assume you are using `0.88.0-SNAPSHOT` ? We are currently looking into some issues with our distribution so the SNAPSHOT might not work correctly.

That said, that `User` implements `UserRealmProxyInterface` is not something you can see in the source code as that is added by our Transform byte manipulation step, so you would have to verify that in the .class file instead.

You should get a warning about it not finding `io/realm/RealmTransform` though. At least that is the error we are seeing currently?

Alternatively. Can you try to to build with `./gradlew clean assemble --info`. That will output a lot of debug information from our bytecode step. If that output is actually there, it looks like something we should look deeper into.

**Comment by cmelchior** [comment]
I am reopening this, just in case it is actually a bigger problem

**Comment by aafa** [comment]
Yes, 0.88.0-SNAPSHOT

**Comment by cmelchior** [comment]
Ok, i'll let you know once we verified everything on our side, because right now I think it is a problem elsewhere /cc @emanuelez

**Comment by aafa** [comment]
Ok, so User.class bytecode should have UserRealmProxyInterface implemented, and I can see that it's not the case.  Meaning that bytecode transformation went not as expected.

I will try that additional debug info step, but I'm building with SBT so that may differ from gradle output.

**Comment by cmelchior** [comment]
Ok, I am not familiar with SBT, but if it doesn't support the Transform API from the Android Gradle toolchain it will definitely not work as we depend on the Transform API step in order to work properly.

**Comment by aafa** [comment]
I've been doing `javac -processor` manually as an additional step between original scala model compile and packaging. I assume it's just not enough. 

Where can I find more information on what actually should be done with bytecode transformation?  Probably I will have to replicate some of that as an SBT task of find a way to get Transform API working.

**Comment by cmelchior** [comment]
We are using the Gradle plugin to register the transformer with Google's API here: https://github.com/realm/realm-java/blob/master/gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy#L46 , but I must confess I haven't looked deeply into how Google integrates it with the build system in the general and if it is possible to trigger this step manually. You would have to do dig into the Googles build tools to discover that.

@emanuelez has been working on it though, so perhaps he has some input that could guide you.

**Comment by cmelchior** [comment]
Hi @aafa 
Unfortunately there doesn't seem to be any easy work-arounds as long as SBT doesn't support the transform API. We have a task for supporting Realm on standard Java (JVM) which means we have to find a way for the bytecode manipulation to happen outside the the Transform API: #2339 

This wil probably also work for SBT, but until then I am closing this issue.

**Comment by aafa** [comment]
Hi @cmelchior,

Actually I managed to get realm working on sbt\scala. I still have some issues ahead but it works.

One thing bothering me is that javassist seems to corrupt particular scala classes while doing read-write on their bytecode. So in order to get this working I've changed this line https://github.com/realm/realm-java/blob/e2ae8a355c4f75a41644bfce572f67f39a04f003/realm-transformer/src/main/groovy/io/realm/transformer/RealmTransformer.groovy#L122  to 

```
            if (BytecodeModifier.useRealmAccessors(ctClass, managedFields, modelClasses)){
                ctClass.writeFile(outputProvider.getContentLocation(
                        'realm', getInputTypes(), getScopes(), Format.DIRECTORY).canonicalPath)
            }
```

where `useRealmAccessors` returns true if Realm is actually modifying  `ctClass`.


So here is my question which I should probably address to @emanuelez \ @zaki50 if I may -- is it really necessary to write _everything_ to `outputProvider` for gradle plugin to work?
